### PR TITLE
add scikit-image package

### PR DESCRIPTION
**Pull request by nopid**
Combines #861 and #733 to add `scikit-image 0.15.0`

A more recent version (like 0.17.2) would require to update `scipy`

No testing yet.

Closes https://github.com/iodide-project/pyodide/pull/733
Closes https://github.com/iodide-project/pyodide/issues/253

### CODE REVIEW

**Comment by nopid** [comment]
Pywavelets needs more work:
```
>>> import pywt
Error: Traceback (most recent call last):
  File "/lib/python3.8/site-packages/pywt/_cwt.py", line 15, in <module>
    import scipy.fft
ModuleNotFoundError: No module named 'scipy.fft'
```

**Comment by nopid** [comment]
A first test taken from the examples. It passes in Firefox and not in Chrome but I see scipy has some tests failing in Chrome too.

**Comment by nopid** [comment]
`imageio` needs more work, it has some `ctypes` dependencies:
```
>>> import imageio
Error: Traceback (most recent call last):
  File "/lib/python3.8/site-packages/pyodide.py", line 66, in eval_code
    exec(compile(mod, "<exec>", mode="exec"), ns, ns)
  File "<exec>", line 1, in <module>
  File "/lib/python3.8/site-packages/imageio/__init__.py", line 22, in <module>
    from .core import FormatManager, RETURN_BYTES
  File "/lib/python3.8/site-packages/imageio/core/__init__.py", line 13, in <module>
    from .findlib import load_lib
  File "/lib/python3.8/site-packages/imageio/core/findlib.py", line 10, in <module>
    import ctypes
ModuleNotFoundError: No module named 'ctypes'
```

**Comment by nopid** [comment]
`imageio` defines a thin wrapper for the `freeimage` library. Without `ctypes` it would require to rewrite it.

Hopefully, `imageio` is used by `scikit-image` as one io plugin among others to read and write images (we already have `pillow` working). We can desactivate this plugin and remove the `imageio` depend to get a working `scikit-image`.
```
[imageio]
description = Image reading via the ImageIO Library
provides = imread, imsave
```

**Comment by nopid** [comment]
`tifffile` needs work, it requires either `lzma` standard Python lib or `imagecodecs` (which has too much requirements):
```
>>> import tifffile
Error: Traceback (most recent call last):
  File "/lib/python3.8/site-packages/pyodide.py", line 66, in eval_code
    exec(compile(mod, "<exec>", mode="exec"), ns, ns)
  File "<exec>", line 1, in <module>
  File "/lib/python3.8/site-packages/tifffile/__init__.py", line 3, in <module>
    from .tifffile import __doc__, __all__, __version__, main
  File "/lib/python3.8/site-packages/tifffile/tifffile.py", line 11315, in <module>
    import lzma
  File "/lib/python3.8/lzma.py", line 27, in <module>
    from _lzma import *
ModuleNotFoundError: No module named '_lzma'
```

**Comment by rth** [comment]
> A more recent version (like 0.17.2) would require to update scipy

Updating scipy (https://github.com/iodide-project/pyodide/issues/549) is somewhat blocked by https://github.com/iodide-project/pyodide/issues/240 as newer versions use blas/lapack more and with the current static linking approach that would translate into even large package size.

>  I see scipy has some tests failing in Chrome too.

yes, it happens occasionally

> We can desactivate this plugin and remove the imageio depend to get a working scikit-image

+1 to desactivate imageio

> tifffile needs work, it requires either lzma standard Python lib

Is there a way to disable tiff file support for now (and not use that library)? Otherwise we could try to build CPython with lzma (but it would increase build time for the core pyodide package).

**Comment by nopid** [comment]
Sure we can desactivate `tifffile` also. The io plugin system of `scikit-image` is quite flexible.

File: packages/scikit-image/patches/no-openmp.patch [hunk 1]
```diff
@@ -0,0 +1,10 @@
+--- a/setup.py	2019-04-01 22:30:11.000000000 +0000
++++ b/setup.py	2020-12-16 06:21:52.600722138 +0000
+@@ -67,6 +67,7 @@
```
Comment (rth) [inline packages/scikit-image/patches/no-openmp.patch:3]: Please add a comment indicating that this patch would no longer be necessary with scikit-image 0.18 (which includes https://github.com/scikit-image/scikit-image/pull/5059)

File: packages/scikit-image/patches/unwrap.patch [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ 
+-cdef extern from *:
+-  void unwrap3D(double* wrapped_volume,
+-                double* unwrapped_volume,
+-                unsigned char* input_mask,
+-                int image_width, int image_height, int volume_depth,
+-                int wrap_around_x, int wrap_around_y, int wrap_around_z,
+-                char use_seed, unsigned int seed) nogil
++
++cdef extern from "unwrap_3d_ljmu.h":
```
Comment (rth) [inline packages/scikit-image/patches/unwrap.patch:46]: Is it something that could be contributed upstream?

Where are they importing it from if they don't have this header file?

Comment (nopid) [reply packages/scikit-image/patches/unwrap.patch:46]: This patch comes directly from a PR upstream.

Comment (rth) [reply packages/scikit-image/patches/unwrap.patch:46]: Then please mention it as a comment inside the patch

**Comment by nopid** [comment]
Thank you @rth for the reviews.

Still missing tests in every packages.

**Comment by rth** [comment]
It looks like `bzip2-1.0.2.tar.gz` can no longer be downloaded from the `ftp://sources.redhat.com/pub` FTP (now that page appears to be password protected). We would need to change that URL to fix CI (in a separate PR).

**Comment by nopid** [comment]
Maybe debian archive would be a reliable place to fetch it : http://archive.debian.org/debian/pool/main/b/bzip2/bzip2_1.0.2.orig.tar.gz

**Comment by nopid** [comment]
rebased to benefit from bzip2 update

**Comment by nopid** [comment]
Simple tests added to every new package by taking some examples from the documentation of the package and adding assertions.

**Comment by nopid** [comment]
To be linked to #253 and #733.

**Comment by rth** [comment]
Thank you @nopid!

**Comment by nopid** [comment]
@rth I pushed the changelog update on the branch just after your merge.

**Comment by rth** [comment]
OK, thanks, I'll add https://github.com/iodide-project/pyodide/commit/92e571fc5d6236e133808f33134385a18c23dd12 to https://github.com/iodide-project/pyodide/pull/813

**Comment by oeway** [comment]
Congrats and thanks to @nopid and @rth! Really happy to see scikit-image is now in pyodide.

**Comment by rth** [comment]
yes, it's great to see it merged! Note however that due to current size of the scipy package it will take a while to load and use a lot of memory. That situation can be improved via either https://github.com/iodide-project/pyodide/issues/240 or https://github.com/iodide-project/pyodide/issues/874

**Comment by oeway** [comment]
> yes, it's great to see it merged! Note however that due to current size of the scipy package it will take a while to load and use a lot of memory. That situation can be improved via either #240 or #874

Thanks! Good to know, and I will wait for the released version to actually test it, looking forward to the next release!

Also, I actually use `imageio` a lot for reading and writing images and I just sent [an issue ](https://github.com/imageio/imageio/issues/580) in the `imageio` repo to see if we can mask out the ctypes part.

**Pull request merged**